libcontainer/nsenter/cloned_binary.c

```diff
@@ -415,32 +514,51 @@ static int try_bindfd(void)
 	close(fd);
 
 	/*
-	 * For obvious reasons this won't work in rootless mode because we haven't
-	 * created a userns+mntns -- but getting that to work will be a bit
-	 * complicated and it's only worth doing if someone actually needs it.
+	 * Daemons such as systemd and udisks2 watch /proc/self/mountinfo and
+	 * re-parse it on every change, which gets expensive when the mount table
+	 * is large and/or changes frequently. Perform the mount operations in a
+	 * new, private mount namespace so as not to wake up those processes
+	 * every time we nsexec into a container. We clone a child process into
+	 * a new mount namespace to do the dirty work so the side effects of
+	 * unsharing the mount namespace do not leak into the current process.
 	 */
-	ret = -EPERM;
-	if (mount("/proc/self/exe", template, "", MS_BIND, "") < 0)
-		goto out;
-	if (mount("", template, "", MS_REMOUNT | MS_BIND | MS_RDONLY, "") < 0)
-		goto out_umount;
-
-	/* Get read-only handle that we're sure can't be made read-write. */
-	ret = open(template, O_PATH | O_CLOEXEC);
+	int sock[2];
+	if (socketpair(AF_LOCAL, SOCK_STREAM, 0, sock) < 0)
+		goto cleanup_unlink;
+
+	struct bindfd_child_args args = {
+		.sockfd = sock[0],
+		.mount_target = template,
+	};
+	int cpid = spawn_bindfd_child(&args);
+	close(sock[0]);
+	if (cpid < 0)
+		goto cleanup_socketpair;
+
+	int wstatus = 0;
+	if (waitpid(cpid, &wstatus, __WCLONE) < 0)
+		bail("error waiting for bindfd child process to exit");
+	if (WIFEXITED(wstatus)) {
+		if ((errno = WEXITSTATUS(wstatus)))
+			goto cleanup_socketpair;
+	} else if (WIFSIGNALED(wstatus)) {
+		int sig = WTERMSIG(wstatus);
+		bail("bindfd child process terminated by signal %d (%s)", sig, strsignal(sig));
+	} else {
+		/* Should never happen... */
+		bail("unexpected waitpid() status for bindfd child process: 0x%x", wstatus);
+	}
 
-out_umount:
-	/*
-	 * Make sure the MNT_DETACH works, otherwise we could get remounted
-	 * read-write and that would be quite bad (the fd would be made read-write
-	 * too, invalidating the protection).
-	 */
-	if (umount2(template, MNT_DETACH) < 0) {
-		if (ret >= 0)
-			close(ret);
-		ret = -ENOTRECOVERABLE;
+	ret = receive_fd(sock[1]);
+	if (fcntl(ret, F_SETFD, O_PATH | O_CLOEXEC) < 0) {
```

Comment:
Should be `FD_CLOEXEC` (which `!= O_CLOEXEC`) and `O_PATH` here doesn't make sense. Also I'm not sure it makes sense to do this anyway -- we know we set `O_CLOEXEC` when we sent it and in general setting `O_CLOEXEC` after the fact is not safe.

I'm a bit surprised you don't get an error from this...

Reply:
@corhere ^^^

Reply:
> Also I'm not sure it makes sense to do this anyway -- we know we set `O_CLOEXEC` when we sent it

I don't think the `O_CLOEXEC` flag of the sending file descriptor is relevant.

> [Semantically, [the SCM_RIGHTS] operation is equivalent to duplicating (`dup(2)`) a file descriptor into the file descriptor table of another process.](https://man7.org/linux/man-pages/man7/unix.7.html#:~:text=Semantically%2C%20this%20operation%20is%20equivalent%20to%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20duplicating%20(dup(2))%20a%20file%20descriptor%20into%20the%20file%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20descriptor%20table%20of%20another%20process.)

> The `dup()` system call allocates a new file descriptor that refers to the same open file description as the descriptor `oldfd`. [...] [The two file descriptors do not share file descriptor flags (the close-on-exec flag). The close-on-exec flag (`FD_CLOEXEC`; see `fcntl(2)`) for the duplicate descriptor is off.](https://man7.org/linux/man-pages/man2/dup.2.html#:~:text=The%20two%20file%20descriptors%20do%20not%20share%20file%20descriptor%20flags%20(the%0A%20%20%20%20%20%20%20close%2Don%2Dexec%20flag).%20%20The%20close%2Don%2Dexec%20flag%20(FD_CLOEXEC%3B%20see%0A%20%20%20%20%20%20%20fcntl(2))%20for%20the%20duplicate%20descriptor%20is%20off.)

That being said, `recvmsg(2)` takes a flag `MSG_CMSG_CLOEXEC` which atomically sets `FD_CLOEXEC` on all received file descriptors. I will set that flag and drop the `fcntl` call.

Reply:
Yeah, looking at [`scm_detach_fds`](https://elixir.bootlin.com/linux/v6.0/source/net/core/scm.c#L297) the receive is done in a context where it can't even know if the original file descriptor was CLOEXEC. But yes, we should be setting `MSG_CMG_CLOEXEC` in `msg_flags`.